src/network/lan.cpp

```diff
@@ -0,0 +1,335 @@
+/*
+Copyright (C) 2016 proller <proler@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
+
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Freeminer  is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "network/lan.h"
+#include <cstdint>
+#include "convert_json.h"
+#include "socket.h"
+#include "log.h"
+#include "settings.h"
+#include "version.h"
+#include "networkprotocol.h"
+#include "server/serverlist.h"
+#include "debug.h"
+#include "json/json.h"
+#include <mutex>
+#include <shared_mutex>
+#include "porting.h"
+#include "threading/thread.h"
+#include "network/address.h"
+
+//copypaste from ../socket.cpp
+#ifdef _WIN32
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+// Without this some of the network functions are not found on mingw
+#ifndef _WIN32_WINNT
+#define _WIN32_WINNT 0x0501
+#endif
+#include <windows.h>
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#define LAST_SOCKET_ERR() WSAGetLastError()
+typedef SOCKET socket_t;
+typedef int socklen_t;
+#else
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <arpa/inet.h>
+
+#ifndef __ANDROID__
+#include <ifaddrs.h>
+#define HAVE_IFADDRS 1
+#endif
+
+#define LAST_SOCKET_ERR() (errno)
+typedef int socket_t;
+#endif
+
+const char* adv_multicast_addr = "224.1.1.1";
```

Comment:
I didn't know that address was in the reserved range. That address wasn't chosen for any particular reason. What should the address be?

Reply:
probably something in the local subnetwork? @nerzhul might be able to assist more, I don't know how this works.
If not local subnet than AD-HOC might be a good choice, in any case reserved ranges in general are meant to be left unused, but maybe this one is an exception.

Reply:
@nerzhul Would the address 224.0.0.24 be acceptable? This address is in the local subnetwork because it is my understanding that this kind of discovery doesn't need to be routed over the internet because we have curl for that. This particular address was chosen at random in that range sense I don't think it matters as long as it is in that range.

Reply:
the first one is fine. i'd say it should be a parameter in case of conflict for some people networks. 224.1.x is reserved and mustn't be used. We should use a non routable range for this feature